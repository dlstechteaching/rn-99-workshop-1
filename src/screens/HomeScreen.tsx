import { useNavigation } from '@react-navigation/native';
import React from 'react';
import { FlatList, SafeAreaView, StyleSheet  } from 'react-native';
import { MyButton } from '../components/common/MyButton';
import { MyTitle } from '../components/common/MyTitle';
import { Navigation } from '../navigation/Navigation';
import { HomeStackScreenProps } from '../navigation/types';
import { dataForOnboardingScreen } from './hardcoded_data';

export const HomeScreen : React.VFC = () => {

	const { navigate } = useNavigation<HomeStackScreenProps['navigation']>();
	const DATA = [
		{ id: '0', label: 'Screen a', screenName: Navigation.A },
		{ id: '1', label: 'Screen b', screenName: Navigation.B },
		{ id: '2', label: 'Screen c', screenName: Navigation.C },
		{ id: '3', label: 'Onboarding', screenName: Navigation.Onboarding, params: { stepIndex: 0, data: dataForOnboardingScreen} },
	];
    
	return (
		<SafeAreaView style={styles.container}>
			<MyTitle label={'Workshop #1'} />
			<FlatList
				style={{ padding: 20 }}
				data={DATA}
				renderItem={({ item }) => 
					<MyButton 
						label={item.label} 
						onPress={() => navigate(item.screenName, item.params)} 
					/>
				}
				keyExtractor={(item) => item.id}
			/>
		</SafeAreaView>
	);
};

const styles = StyleSheet.create({
	container: {
		flex: 1,
		marginTop: 0,
		padding: 20
	},
});