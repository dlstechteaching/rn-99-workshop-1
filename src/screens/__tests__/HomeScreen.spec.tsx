import { useNavigation } from '@react-navigation/native';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import { HomeScreen } from '../HomeScreen';

jest.mock('@react-navigation/native');

describe('HomeScreen tests', () => {
	let component: ShallowWrapper;
	const navigationMock = {
		navigate: jest.fn(),
	};

	const itemMock = {
		label: 'test',
		screenName: 'screen',
	};

	beforeEach(() => {
		(useNavigation as jest.Mock).mockReturnValue(navigationMock);
	});

	it('should render properly', () => {
		givenComponent();
		thenItRendersProperly();
	});

	it('should navigato to specific screen', () => {
		givenComponent();
		whenFirstMyButtonIsClicked();
		thenItNavigates();
	});

	function givenComponent() {
		component = shallow(<HomeScreen />);
	}

	function whenFirstMyButtonIsClicked() {
		component
			.find('FlatList')
			.props()
			// @ts-ignore
			.keyExtractor({ id: 1 }, 0);

		const firstButton = shallow(
			component
				.find('FlatList')
				.props()
				// @ts-ignore
				.renderItem({ item: itemMock }),
		);
		// @ts-ignore
		firstButton.props().onPress();
	}

	function thenItRendersProperly() {
		expect(component).toMatchSnapshot();
	}

	function thenItNavigates() {
		expect(navigationMock.navigate).toHaveBeenCalled();
	}
});
