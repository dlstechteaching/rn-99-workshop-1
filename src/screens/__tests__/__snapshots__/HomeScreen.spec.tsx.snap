// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen tests should render properly 1`] = `
<ForwardRef(SafeAreaView)
  style={
    Object {
      "flex": 1,
      "marginTop": 0,
      "padding": 20,
    }
  }
>
  <MyTitle
    label="Workshop #1"
  />
  <FlatList
    data={
      Array [
        Object {
          "id": "0",
          "label": "Screen a",
          "screenName": "app.ex.a",
        },
        Object {
          "id": "1",
          "label": "Screen b",
          "screenName": "app.ex.b",
        },
        Object {
          "id": "2",
          "label": "Screen c",
          "screenName": "app.ex.c",
        },
        Object {
          "id": "3",
          "label": "Onboarding",
          "params": Object {
            "data": Array [
              Object {
                "desc": "We can help you to be a better version of yourself",
                "image": 1,
                "title": "Welcome to monumental habits",
              },
              Object {
                "desc": "We can help you to be a better version of yourself",
                "image": 1,
                "title": "Create new habit easily",
              },
              Object {
                "desc": "We can help you to be a better version of yourself",
                "image": 1,
                "title": "Keep track of your progress",
              },
              Object {
                "desc": "We can help you to be a better version of yourself",
                "image": 1,
                "title": "Join a supportive community",
              },
            ],
            "stepIndex": 0,
          },
          "screenName": "app.onboarding",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    style={
      Object {
        "padding": 20,
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</ForwardRef(SafeAreaView)>
`;
